pkg/tecdsa/signing/states.go

```diff
@@ -997,3 +793,115 @@ func (fs *finalizationState) MemberIndex() group.MemberIndex {
 func (fs *finalizationState) result() *Result {
 	return fs.member.Result()
 }
+
+// stateAction represents an asynchronous action performed in the given
+// protocol state.
+type stateAction struct {
+	mutex   sync.RWMutex
+	running bool
+	done    bool
+	err     error
+}
+
+// run triggers the action goroutine. Can be called only once. Successive
+// calls do nothing.
+func (sa *stateAction) run(actionFn func() error) {
+	sa.mutex.Lock()
+	defer sa.mutex.Unlock()
+
+	if sa.running || sa.done {
+		return
+	}
+
+	sa.running = true
+
+	go func() {
+		err := actionFn()
+
+		sa.mutex.Lock()
+
+		sa.running = false
+		sa.done = true
+		sa.err = err
+
+		sa.mutex.Unlock()
+	}()
+}
+
+// isDone returns whether the state's action is done.
+func (sa *stateAction) isDone() bool {
+	sa.mutex.RLock()
+	defer sa.mutex.RUnlock()
+	return sa.done
+}
+
+// error returns the state's action error if any. Calling this function makes
+// sense only when the action is done, i.e. the isDone function returns true.
+// After this function returns a non-nil error, successive calls return the
+// same error.
+func (sa *stateAction) error() error {
+	sa.mutex.RLock()
+	defer sa.mutex.RUnlock()
+	return sa.err
+}
+
+// sendMessage sends the state's message using the provided broadcast channel.
+func sendMessage(
+	ctx context.Context,
+	channel net.BroadcastChannel,
+	message net.TaggedMarshaler,
+) error {
+	// The channel.Send function sets up retransmissions and expects a
+	// unique context to properly differentiate the sending routine among
+	// others. The context received as argument of the Initiate function
+	// cannot be used directly because it is global for the entire
+	// state machine. Because of that, channel.Send must be called with
+	// a child context.
+	sendCtx := context.WithValue(ctx, "message", message.Type())
```

Comment:
Why do we use `WithValue`? Is it to trick the `ticker.go` to distinguish contexts and not treat contexts for two different states as equal?

Reply:
We don't have a simple way to create a context copy in Go apart wrapping it with `context.WithValue`. All other functions like `WithCancel` or `WithDeadline` produce a cancel function that must be managed somehow. We can't just ignore it as, in theory, we don't have a guarantee that the parent context is always canceled (that will cause context leaks). On the other hand, we cannot setup a proper call to the cancel function at this level. This is why the `WithValue` trick is a good option here, especially if we treat this solution as a temporary one until we fix the problem directly in `ticker.go`.

Reply:
Addressed in #3383.